### PR TITLE
Workaround restart rabbitmq pods during patching #1395

### PR DESCRIPTION
**Pull request by ar3ndt**

### CODE REVIEW

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 1]
```diff
@@ -45,3 +45,28 @@
       when:
         - not rabbit_mq_namespace.stdout == ""
         - not image_registry_address in rabbit_mq_repository.stdout
+    - name: upgrade-master | Get rabbitmq pod names
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:51]: I see that there is no pipe so pipefail is not needed. I would change `shell` module to `command`.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 2]
```diff
@@ -45,3 +45,28 @@
...
+      changed_when: false
+      register: rabbit_mq_pod_names
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+    - name: upgrade-master | Delete rabbitmq pods
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:63]: Same here.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 3]
```diff
@@ -3,3 +3,70 @@
...
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+    - name: upgrade-master | Patch rabbitmq to use {{ image_registry_address }}
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
+        set -o pipefail &&
+        kubectl patch statefulsets.apps {{ rabbit_mq_ss_name.stdout }} -n {{ rabbit_mq_namespace.stdout }} --patch '{"spec": {"template": { "spec": { "containers": [ { "image": "{{ image_registry_address }}/{{ rabbit_mq_repository.stdout }}", "name": "{{ rabbit_mq_ss_name.stdout }}" }]}}}}'
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:46]: I don't think `shell` is needed here, I would change it to `command` and remove `set -o pipe`.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 4]
```diff
@@ -3,3 +3,70 @@
...
+        awk '{print $1}'
+      changed_when: false
+      register: rabbit_mq_namespace
+      args:
+        executable: /bin/bash
+    - name: upgrade-master | Get rabbitmq image
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
+        kubectl get statefulsets.apps {{ rabbit_mq_ss_name.stdout }} -n {{ rabbit_mq_namespace.stdout }} -o=jsonpath='{$.spec.template.spec.containers[:1].image}'
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:34]: I don't think `shell` is needed here, I would change it to `command`

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 5]
```diff
@@ -3,3 +3,70 @@
...
+      register: rabbit_mq_ss_name
+      args:
+        executable: /bin/bash
+    - name: upgrade-master | Get rabbitmq namespace
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
+        kubectl get statefulsets.apps --all-namespaces -o=jsonpath='{range .items[*]}{.metadata.namespace}{"\t"}{.spec.template.spec.containers[].image}{"\n"}{end}'|
+        grep -i rabbitmq |
+        awk '{print $1}'
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:25]: Same here

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 6]
```diff
@@ -3,3 +3,70 @@
...
     image_regexp: 'rabbitmq:.*'
+- name: Change rabbitmq stateful set to use {{ image_registry_address }}
+  block:
+    - name: upgrade-master | Get rabbitmq statefulset name
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
+        kubectl get statefulsets.apps --all-namespaces -o=jsonpath='{range .items[*]}{.metadata.name}{"\t"}{.spec.template.spec.containers[].image}{"\n"}{end}'|
+        grep -i rabbitmq |
+        awk '{print $1}'
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:14]: I would get rid of additional pipe here, pls consider to use here only awk (eg. awk '/rabbitmq/ {print $1}'

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 7]
```diff
@@ -3,3 +3,70 @@
...
+        kubectl patch statefulsets.apps {{ rabbit_mq_ss_name.stdout }} -n {{ rabbit_mq_namespace.stdout }} --patch '{"spec": {"template": { "spec": { "containers": [ { "image": "{{ image_registry_address }}/{{ rabbit_mq_repository.stdout }}", "name": "{{ rabbit_mq_ss_name.stdout }}" }]}}}}'
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+        - not image_registry_address in rabbit_mq_repository.stdout
+    - name: upgrade-master | Get rabbitmq pod names
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: kubectl get pod -n {{ rabbit_mq_namespace.stdout }} -o=jsonpath='{range .items[*]}{.metadata.name}{"\n"}'
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:55]: Same here

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 8]
```diff
@@ -3,3 +3,70 @@
...
+      changed_when: false
+      register: rabbit_mq_pod_names
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+    - name: upgrade-master | Delete rabbitmq pods
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: kubectl delete pod --namespace {{ rabbit_mq_namespace.stdout }} {{ item }}
```
Comment (rafzei) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:65]: Same here

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 9]
```diff
@@ -3,3 +3,70 @@
   include_tasks: utils/patch-statefulset.yml
   vars:
     image_regexp: 'rabbitmq:.*'
+- name: Change rabbitmq stateful set to use {{ image_registry_address }}
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:6]: I think this code is not needed at all, please take a look at `utils/patch-statefulset.yml`. We refactored it lately. What I can see here is some redundant code from before the refactor and it does not take into account init containers for example (which was and is a bug tbh).
Also `KUBECONFIG: /home/{{ admin_user.name }}/.kube/config` is no longer valid, we decided to remove to completely as it does not bring any benefits. We started using `/etc/kubernetes/admin.conf` exclusively and we want to continue that. :+1:

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 10]
```diff
@@ -3,3 +3,70 @@
...
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
+      shell: |-
+        set -o pipefail &&
+        kubectl patch statefulsets.apps {{ rabbit_mq_ss_name.stdout }} -n {{ rabbit_mq_namespace.stdout }} --patch '{"spec": {"template": { "spec": { "containers": [ { "image": "{{ image_registry_address }}/{{ rabbit_mq_repository.stdout }}", "name": "{{ rabbit_mq_ss_name.stdout }}" }]}}}}'
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+        - not image_registry_address in rabbit_mq_repository.stdout
+    - name: upgrade-master | Get rabbitmq pod names
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:52]: As I understand here is the place where the valid code starts.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 11]
```diff
@@ -3,3 +3,70 @@
...
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:54]:
```suggestion
        KUBECONFIG: /etc/kubernetes/admin.conf
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 12]
```diff
@@ -3,3 +3,70 @@
...
+      shell: kubectl get pod -n {{ rabbit_mq_namespace.stdout }} -o=jsonpath='{range .items[*]}{.metadata.name}{"\n"}'
+      changed_when: false
+      register: rabbit_mq_pod_names
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+    - name: upgrade-master | Delete rabbitmq pods
+      environment:
+        KUBECONFIG: /home/{{ admin_user.name }}/.kube/config
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:64]:
```suggestion
        KUBECONFIG: /etc/kubernetes/admin.conf
```

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml [hunk 13]
```diff
@@ -3,3 +3,70 @@
...
+      shell: |-
+        kubectl get statefulsets.apps {{ rabbit_mq_ss_name.stdout }} -n {{ rabbit_mq_namespace.stdout }} -o=jsonpath='{$.spec.template.spec.containers[:1].image}'
+      changed_when: false
+      register: rabbit_mq_repository
+      args:
+        executable: /bin/bash
+      when:
+        - not rabbit_mq_namespace.stdout == ""
+    - name: upgrade-master | Patch rabbitmq to use {{ image_registry_address }}
```
Comment (sk4zuzu) [inline core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:41]: `name: upgrade-master | Patch rabbitmq to use` this whole task is redundant please remove it.

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml (hunk 5 above)
Comment (ar3ndt) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml:25]: pipe does not work with command

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml (hunk 9 above)
Comment (ar3ndt) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml]: which part do you mean is legacy method?

File: core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml (hunk 5 above)
Comment (rafzei) [reply core/src/epicli/data/common/ansible/playbooks/roles/upgrade/tasks/kubernetes/reconfigure-rabbitmq-app.yml]: yes, the same hare was about using only awk

**Pull request merged**